rcl_yaml_param_parser/src/parse.c

```diff
@@ -809,18 +801,37 @@ rcutils_ret_t find_parameter(
       return RCUTILS_RET_OK;
     }
   }
-  // Parameter not found, add it.
   rcutils_allocator_t allocator = param_st->allocator;
+  // Reallocate if necessary
+  bool reallocate = false;
+  rcutils_ret_t ret;
+  if (node_param_st->num_params >= node_param_st->capacity_params) {
+    if (RCUTILS_RET_OK != node_params_reallocate(
+        node_param_st, allocator, node_param_st->capacity_params * 2))
+    {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+    reallocate = true;
+  }
+  // Parameter not found, add it.
   if (NULL != node_param_st->parameter_names[*parameter_idx]) {
     param_st->allocator.deallocate(
       node_param_st->parameter_names[*parameter_idx], param_st->allocator.state);
   }
   node_param_st->parameter_names[*parameter_idx] = rcutils_strdup(parameter_name, allocator);
   if (NULL == node_param_st->parameter_names[*parameter_idx]) {
-    return RCUTILS_RET_BAD_ALLOC;
+    ret = RCUTILS_RET_BAD_ALLOC;
+    goto clean;
   }
   node_param_st->num_params++;
   return RCUTILS_RET_OK;
+
+clean:
+  if (reallocate) {
+    rcl_yaml_node_params_fini(node_param_st, allocator);
+  }
+  return ret;
```

Comment:
I think this is a bit strange.  That is, if we fail because we couldn't reallocate, we *don't* clean up the memory, but if we fail for another reason later on, we *do* clean up the memory.  How is a caller of this function supposed to react to a failure?  Call fini, or not call fini?  I don't think they can know.  I think we need to think about the post-conditions of this function in both success and failure cases, document it in the header file, and then implement whatever we decide there.

(this same comment applies to `find_node` below)

Reply:
> but if we fail for another reason later on, we do clean up the memory. How is a caller of this function supposed to react to a failure? Call fini, or not call fini?

Call `fini` is the responsibility inside the function. Actually, I think it's the same pattern as
```
function:
   a1 = allocate
   b1 = reallocate
   c1 = do_some
   if c1 is error, goto clean

   return true   
clean:
   deallocate(b1)
   deallocate(a1)
```

Reply:
I found there is a problem,

https://github.com/ros2/rcl/blob/c1b3ff72d3d106e322661db3a2702b476eedfe27/rcl_yaml_param_parser/src/parser.c#L167

can only be called just once, and a function to call `rcl_yaml_node_struct_fini` must assign the `params_st` with NULL immediately, so I can't use this function in `clean` location because of some allocate failure.